### ISSUE TITLE
sourceFile duplication for TypedRule

### ISSUE THREAD

**Issue by andy-hanson** [palantir/tslint#1969]
### Bug Report

- __TSLint version__: `master`
- __TypeScript version__: `2.1.4` (the one in `package.json`)
- __Running TSLint via__: `./bin/tslint --test test/rules/my-rule`

#### TypeScript code ~~being linted~~

```ts
export class Rule extends Lint.Rules.TypedRule {
    /* tslint:disable:object-literal-sort-keys */
    public static metadata: Lint.IRuleMetadata = {
        ruleName: "my-rule",
        description: "",
        optionsDescription: "Not configurable.",
        options: null,
        optionExamples: ["true"],
        type: "functionality",
        typescriptOnly: false,
    };
    /* tslint:enable:object-literal-sort-keys */

    public applyWithProgram(sourceFile: ts.SourceFile, langSvc: ts.LanguageService): Lint.RuleFailure[] {
        console.log(sourceFile === langSvc.getProgram().getSourceFile(sourceFile.fileName));
        return [];
    }
}
```

#### Actual behavior

Prints `false`.

#### Expected behavior

Prints `true`. The source file used for linting should be the same as the one used for the program.
I want to write a rule that uses the type checker to get information about nodes as I recurse. I can't do that if it uses a different AST.

**Comment by nchen63** [comment]
Is the AST different or are the sourceFile objects different references?

**Comment by andy-hanson** [comment]
Looks like they're completely separate. Added this to the top of `applyWithProgram` in `noForInArrayRule.ts`:
```ts
const langSvcSourceFile = langSvc.getProgram().getSourceFile(sourceFile.fileName);
console.log(sourceFile === langSvcSourceFile);
console.log(sourceFile.statements[0] === langSvcSourceFile.statements[0]);
process.exit(0);
```

Both lines print `false`.

**Comment by andy-ms** [comment]
Since it's the same file being parsed twice, the two separate source files should have identical contents.
